ambry-protocol/src/main/java/com/github/ambry/protocol/AmbryRequests.java

```diff
@@ -1030,6 +1039,271 @@ private Pair<ServerErrorCode, GetResponse> getBlobFromRemoteReplica(ReplicateBlo
+      logger.info("ReplicateBlobRequest invalid life version {}", replicateBlobRequest);
+      completeReplicateRequest(request, replicateBlobRequest, ServerErrorCode.Bad_Request, startProcessTime);
+      return;
+    }
+
+    // this replica itself is the source replica. Don't need do anything. Return success status.
+    DataNodeId remoteDataNode = clusterMap.getDataNodeId(remoteHostName, remoteHostPort);
+    if (remoteDataNode != null && remoteDataNode.equals(currentNode)) {
+      logger.error("ReplicateBlobRequest this is the source replica, return immediately. {}", replicateBlobRequest);
+      completeReplicateRequest(request, replicateBlobRequest, ServerErrorCode.No_Error, startProcessTime);
+      return;
+    }
+
+    // check if local store has the key already
+    ServerErrorCode errorCode = ServerErrorCode.No_Error;
+    MessageInfo localMessageInfo = null;
+    try {
+      StoreKey convertedKey = getConvertedStoreKeys(Collections.singletonList(blobId)).get(0);
+      if (!(convertedKey).equals(blobId)) {
+        logger.info("ReplicateBlobRequest replicated Blob {} convert to local key {}", blobId, (BlobId) convertedKey);
+        blobId = (BlobId) convertedKey;
+      }
+      Store store = storeManager.getStore(blobId.getPartition());
+      localMessageInfo = store.findKey(convertedKey);
+      logger.trace("ReplicateBlobRequest replicated Blob {}, local Store has the Key", blobId);
+    } catch (StoreException e) {
+      // store.findKey throws e.getErrorCode() == StoreErrorCodes.ID_Not_Found if it doesn't exist.
+      if (e.getErrorCode() == StoreErrorCodes.ID_Not_Found) {
+        errorCode = ServerErrorCode.No_Error;
+      } else {
+        logger.error("ReplicateBlobRequest store.findKey throw exception {} on Blob {},", e.getErrorCode(), blobId);
+        errorCode = ErrorMapping.getStoreErrorMapping(e.getErrorCode());
+      }
+    } catch (Exception e) {
+      // getConvertedStoreKeys which may throw Exception
+      logger.error("ReplicateBlobRequest getConvertedStoreKeys unknown exception to replicate {} of {}", blobId,
+          replicateBlobRequest, e);
+      errorCode = ServerErrorCode.Unknown_Error;
+    }
+    if (errorCode != ServerErrorCode.No_Error) {
+      completeReplicateRequest(request, replicateBlobRequest, errorCode, startProcessTime);
+      return;
+    }
+
+    if (replicateBlobRequest.getOperationType() == RequestOrResponseType.TtlUpdateRequest) {
+      // Repair the PutBlob if local store doesn't have the blob when repair type is RequestOrResponseType.TtlUpdateRequest
+      if (localMessageInfo == null) {
+        Pair<ServerErrorCode, GetResponse> remoteGetResult = getBlobFromRemoteReplica(replicateBlobRequest);
+        errorCode = remoteGetResult.getFirst();
+        GetResponse remoteGetResponse = remoteGetResult.getSecond();
+        if (errorCode == ServerErrorCode.No_Error) {
+          errorCode = repairPutBlob(replicateBlobRequest, blobId, remoteGetResponse);
+        }
+      }
```

Comment:
I think we should repair the put on both operation types. 

if we switch to V2 in frontend, and now this is a on demand replication for delete, we should try to get the blob content before applying the delete. In this way, we would be consistent with the source replica.

Reply:
Now for ReplicateDelete,
will replicate PutBlob, and then tombstone if Put doesn't exist.